src/StepperMotorReadback.cc

```diff
@@ -7,118 +7,172 @@
+        try {
+          // FIXME This is only to evaluate the timer
+          if(!execTimer.isInitialized()) {
+            execTimer.initializeMeasurement();
+          }
+          else {
+            execTimer.measureIterativeMean();
+          }
+          auto ct = std::chrono::duration_cast<std::chrono::microseconds>(execTimer.getMeasurementResult());
+          actualCycleTime = static_cast<float>(ct.count()) / 1000.f;
+
+          receiveTimer.initializeMeasurement();
+
+          readback();
+          if(_motor->get()->hasHWReferenceSwitches()) {
+            readEndSwitchData();
+          }
+
+          receiveTimer.measureOnce();
+          auto rt = std::chrono::duration_cast<std::chrono::microseconds>(receiveTimer.getMeasurementResult());
+          actualReceiveTime = static_cast<float>(rt.count()) / 1000.f;
+
+          // We have aparently communicated successfully. Reset the error counter
+          spiErrorCounter = 0;
+        }
+        catch(mtca4u::MotorDriverException& e) {
+          if((e.getID() == mtca4u::MotorDriverException::SPI_ERROR ||
+                 e.getID() == mtca4u::MotorDriverException::SPI_TIMEOUT)) {
+            if(spiErrorCounter > SPI_RETRY_COUNT) {
+              std::cerr << "motor device failed: " << _motor->toString() << std::endl;
+              incrementDataFaultCounter();
+              _motor->close();
+              deviceError.status = static_cast<int32_t>(StatusOutput::Status::FAULT);
+              deviceError.message = e.what();
+              deviceError.setCurrentVersionNumber({});
+              deviceError.writeAll();
+            }
+            else {
+              spiErrorCounter++;
+            }
+          }
+          else {
+            std::rethrow_exception(std::current_exception());
+          }
+        }
       }
 
-      receiveTimer.measureOnce();
-      auto rt = std::chrono::duration_cast<std::chrono::microseconds>(receiveTimer.getMeasurementResult());
-      actualReceiveTime = static_cast<float>(rt.count()) / 1000.f;
-
-      //We now have all data. Write them
+      //We now have all data. Write them, whether or not the read failed. this will set the validity properly.
       writeAll();
 
       //Wait for cyclic trigger
       trigger.read();
+
+      if(not _motor->isOpen()) {
+        try {
+          _motor->renew();
+          std::cerr << "device recovered: " << _motor->toString() << std::endl;
+          decrementDataFaultCounter();
+          deviceError.status = static_cast<int32_t>(StatusOutput::Status::OK);
+          deviceError.message = "";
+          deviceError.setCurrentVersionNumber({});
+          deviceError.writeAll();
+          spiErrorCounter = 0;
+        }
+        catch(mtca4u::MotorDriverException& e) {
+          if(std::string(deviceError.message) != e.what()) {
+            deviceError.message = e.what();
+            deviceError.setCurrentVersionNumber({});
+            deviceError.message.write();
+          }
```

Comment:
Five closing braces are a lot. One could factor stuff out into a function.

Reply:
Just a comment, no action required for the pull request.

Reply:
Fixed.